File: Chap2_LinkedLists/RemoveDups.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;

class Main {
    public static void main(String[] args) {
        Node h = makeLL(new int[] { 1, 2, 3, 4, 5, 6, 6, 6, 6, 6, 7, 8, 9, 1 });
        printLL(h);
        removeDups(h);
        printLL(h);
        h = makeLL(new int[] { 3, 8, 1, 2, 3, 4, 5, 6, 6, 6, 611 });
        printLL(h);
        removeDups(h);
        printLL(h);
        h = makeLL(new int[] { 3, 3, 4, 5, 6, 7, 8, 1, 2, 3, 66, 7, 8 });
        printLL(h);
        removeDupsSpace1(h);
        printLL(h);
    }

    // Make a HashSet as a buffer and traverse LinkedList to check duplicates and
    // remove them
    // time:O(n), n is the length of the linked list (if s.contains is O(1))
    // space:O(n), we build a set which is at most as long as the linked list
    static Node removeDups(Node head) {
        if (head == null) {
            return head;
        }
        Set<Integer> s = new HashSet<>();
        s.add(head.val);
        Node cur = head;
        while (cur.next != null) {
            if (s.contains(cur.next.val)) {
                cur.next = cur.next.next;
            } else {
                s.add(cur.next.val);
                cur = cur.next;
            }
        }
        return head;
    }

    // We don't use extra space to record the duplicate elements
    // We use a runner to check all future elements and skip duplicate ones
    // time:O(n^2)
    // space:O(1)
    static Node removeDupsSpace1(Node head) {
        Node cur = head;
        while (cur != null) {
            Node runner = cur;
            while (runner.next != null) {
                if (runner.next.val == cur.val) {
                    runner.next = runner.next.next;
                } else {
                    runner = runner.next;
                }
            }
            cur = cur.next;
        }
        return head;
    }

    static Node makeLL(int[] arr) {
        Node head = new Node(arr[0]);
        Node cur = head;
        for (int i = 1; i < arr.length; i++) {
            cur.next = new Node(arr[i]);
            cur = cur.next;
        }
        cur.next = null;
        return head;
    }

    static void printLL(Node head) {
        if (head == null) {
            return;
        }
        while (head != null) {
            System.out.print(head.val);
            System.out.print("->");
            head = head.next;
        }
        System.out.print("x\n");
    }
}

class Node {
    int val;
    Node next;

    Node(int v) {
        val = v;
    }
}

File: Chap2_LinkedLists/2.1_RemoveDups/RemoveDups.js
/**
 * @param {number} v
 */
function Node(v) {
    this.val = v;
    this.next = null;
}

/**
 * @param {Node} head
 * @returns {Node}
 * time:O(n), we traverse the linked list with n elements
 * space:O(n), we make a set with potential size n
 */
function removeDups(head) {
    if (!head) {
        return head;
    }
    let s = [head.val];
    let cur = head;
    while (cur.next) {
        if (s.includes(cur.next.val)) {
            cur.next = cur.next.next;
        } else {
            s.push(cur.val);
            cur = cur.next;
        }
    }
    return head;
}

/**
 * param {Node} head
 * @returns {void}
 * we use a runner pointer to check all future nodes
 * time:O(n^2), n is the length of the linked list
 * space:O(1)
 */
function removeDupsSpace1(head) {
    let cur = head;
    while (cur) {
        let runner = cur;
        while (runner.next) {
            if (runner.next.val === cur.val) {
                runner.next = runner.next.next;
            } else {
                runner = runner.next;
            }
        }
        cur = cur.next;
    }
}

function test() {
    let h = makeLL([1, 2, 3, 4, 5, 6, 1, 2]);
    printLL(h);
    removeDups(h);
    printLL(h);
    h = makeLL([1, 2, 3, 4, 5, 6, 1, 2]);
    printLL(h);
    removeDupsSpace1(h);
    printLL(h);
}

/**
 * @param {number[]} arr
 * @returns {Node}
 */
function makeLL(arr) {
    if (!arr || arr.length < 0) {
        return null;
    }
    let head = new Node(arr[0]);
    let cur = head;
    for (let i = 1; i < arr.length; i++) {
        cur.next = new Node(arr[i]);
        cur = cur.next;
    }
    return head;
}

/**
 * @param {Node} h
 * @returns {void}
 */
function printLL(h) {
    if (!h) {
        return;
    }
    let out = "";
    while (h) {
        out += h.val + "->";
        h = h.next;
    }
    console.log(out + "x");
}